src/vs/editor/contrib/rtv/RTVSynthService.ts

```diff
@@ -0,0 +1,327 @@
+	private _boxVars: Set<string> = new Set<string>();
+	private _lineNumber: number;
+	private _rowsValid: boolean[] = [];
+	private _includedTimes: Set<number> = new Set();
+	private _outputVars: string[];
+	private _rows?: TableElement[][];
+	private onBoxContentChanged?: Function;
+
+	constructor(
+		outputVars: string[],
+		lineno: number,
+		boxVars: Set<string>
+	) {
+		this._outputVars = outputVars;
+		this._lineNumber = lineno;
+		this._boxVars = boxVars;
+	}
+
+	get boxEnvs(): any[] {
+		return this._boxEnvs;
+	}
+
+	get includedTimes(): Set<number> {
+		return this._includedTimes;
+	}
+
+	get prevEnvs(): Map<number, any> {
+		return this._prevEnvs!;
+	}
+
+	get varnames(): string[] {
+		return this._outputVars!;
+	}
+
+	bindBoxContentChanged(callback: Function) {
+		this.onBoxContentChanged = callback;
+	}
+
+	_commit(init: boolean = false) {
+		this.onBoxContentChanged!(
+			{
+				'rows': this._rows,
+				'includedTimes': this._includedTimes,
+				'boxEnvs': this._boxEnvs
+			}, init);
+	}
+
+	public updateBoxContent(newEnvs: any, init: boolean = false) {
+		this.updateBoxEnvs(newEnvs);
+		this.updateRowsValid();
+		this._commit(init);
+	}
+
+
+	/**
+	 * Updates `allEnvs` and `prevEnvs`
+	 * @param runResults
+	 * @param includedTimes
+	 */
+	public updateAllEnvs(runResults: any, includedTimes?: Set<number>): void {
+		if (includedTimes) {
+			this._includedTimes = includedTimes;
+		}
+
+		this._allEnvs = [];
+		for (let line in (runResults[2] as { [k: string]: any[]; })) {
+			this._allEnvs = this._allEnvs.concat(runResults[2][line]);
+		}
+
+		this._prevEnvs = new Map<number, any>();
+
+		for (const startEnv of this._allEnvs) {
+			const start = startEnv['time'];
+			let minDelta = 1024 * 1024;
+			let minEnv = undefined;
+
+			for (const env of this._allEnvs) {
+				const time = env['time'];
+				if (time) {
+					const delta = start - time;
+					if (delta > 0 && delta < minDelta) {
+						minDelta = delta;
+						minEnv = env;
+
+						if (delta === 1) {
+							break;
+						}
+					}
+				}
+			}
+
+			if (minEnv) {
+				this._prevEnvs.set(start, minEnv);
+			}
+		}
+	}
+
+	/**
+	 * Updates `boxEnvs' and builds `rows`
+	 * @param newEnvs
+	 */
+	public updateBoxEnvs(newEnvs: any[]) {
+
+		let outVarNames: string[];
+		if (!this._outputVars) {
+			outVarNames = [];
+		} else {
+			outVarNames = this._outputVars!;
+		}
+
+		this._boxEnvs = this.computeEnvs(newEnvs);
+		let envs = this._boxEnvs;
+		let vars = this._boxVars;
+
+		if (this._prevEnvs) {
+			const oldVars = vars;
+			vars = new Set();
+			for (const v of oldVars) {
+				// remove any variables newly defined by the synthsizer
+				let rs = true;
+				if (outVarNames.includes(v)) {
+					for (const env of envs) {
+						const time = env['time'];
+						const prev = this._prevEnvs.get(time);
+						if (prev) {
+							rs = v in prev;
+						}
+					}
+				}
+
+				if (rs) {
+					vars.add(v);
+				}
+			}
+		}
+
+		// Generate header
+		let rows: TableElement[][] = [];
```

Comment:
Are `TableElement`s used outside of creating the HTML view? If not, maybe the view can handle creating these.

Reply:
Not really (the class itself is also used in normal PB code `RTVDisplay.ts`, though). However, the `vname` field depends on `this._prevEnvs` in `RTVSynthModel` as we sometimes need to add the `_in` suffix. Since I now have removed all box state related data from `RTVSynthView`, I think it's better that we keep the building of `TableElement[][]` inside `RTVSynthModel`.